File: 40. SpellOfTwinEnergies/SpellOfTwinEnergies_sol.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int t = sc.nextInt();  

        while (t-- > 0) {
            long c = sc.nextLong();  
            System.out.println(canStabilize(c));
        }

        sc.close();
    }
    static boolean canStabilize(long c) {
        long spellA = 0;
        long spellB = (long) Math.sqrt(c);

        while (spellA <= spellB) {
            long totalPower = spellA * spellA + spellB * spellB;

            if (totalPower == c) {
                return true;   
            } else if (totalPower < c) {
                spellA++;        
            } else {
                spellB--;       
            }
        }

        return false;
    }
}

File: 40. SpellOfTwinEnergies/SpellOfTwinEnergiesSol.cpp
#include <iostream>
#include <cmath>
using namespace std;

// Function to check if c can be written as a^2 + b^2
bool canStabilize(int c) {
    int spellA = 0;
    int spellB = static_cast<int>(sqrt(c));

    while (spellA <= spellB) {
        int totalPower = spellA * spellA + spellB * spellB;

        if (totalPower == c) {
            return true;
        } else if (totalPower < c) {
            spellA++;
        } else {
            spellB--;
        }
    }

    return false;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int t;
    cin >> t;

    while (t--) {
        int c;
        cin >> c;
        cout << (canStabilize(c) ? "true" : "false") << "\n";
    }

    return 0;
}
